File: code/assurance/src/test/java/com/adobe/marketing/mobile/assurance/AssuranceStateManagerTest.java
/*
 * Copyright 2022 Adobe. All rights reserved.
 * This file is licensed to you under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License. You may obtain a copy
 * of the License at http://www.apache.org/licenses/LICENSE-2.0
 * Unless required by applicable law or agreed to in writing, software distributed under
 * the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR REPRESENTATIONS
 * OF ANY KIND, either express or implied. See the License for the specific language
 * governing permissions and limitations under the License.
 */

package com.adobe.marketing.mobile.assurance;

import static org.mockito.ArgumentMatchers.any;
import static org.mockito.ArgumentMatchers.eq;
import static org.mockito.ArgumentMatchers.nullable;
import static org.mockito.Mockito.doAnswer;
import static org.mockito.Mockito.doReturn;
import static org.mockito.Mockito.reset;
import static org.mockito.Mockito.times;
import static org.mockito.Mockito.verify;
import static org.mockito.Mockito.when;

import android.app.Application;
import android.content.Context;
import android.content.SharedPreferences;
import com.adobe.marketing.mobile.Event;
import com.adobe.marketing.mobile.EventSource;
import com.adobe.marketing.mobile.EventType;
import com.adobe.marketing.mobile.ExtensionApi;
import com.adobe.marketing.mobile.SharedStateResolution;
import com.adobe.marketing.mobile.SharedStateResult;
import com.adobe.marketing.mobile.SharedStateStatus;
import com.google.gson.Gson;
import java.util.Collections;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import junit.framework.TestCase;
import org.junit.Test;
import org.mockito.ArgumentCaptor;
import org.mockito.Mock;
import org.mockito.MockitoAnnotations;
import org.mockito.invocation.InvocationOnMock;
import org.mockito.stubbing.Answer;

public class AssuranceStateManagerTest extends TestCase {

    @Mock private ExtensionApi mockApi;

    @Mock private Application mockApplication;

    @Mock SharedPreferences mockSharedPreference;

    @Mock SharedPreferences.Editor mockSharedPreferenceEditor;

    private AssuranceStateManager assuranceStateManager;

    private Map<String, Object> SAMPLE_STATE_DATA;
    private Map<String, Object> SAMPLE_XDM_STATE_DATA;

    public void setUp() throws Exception {
        MockitoAnnotations.openMocks(this);

        SAMPLE_STATE_DATA = new HashMap<String, Object>();
        SAMPLE_STATE_DATA.put("stateKey", "stateValue");

        SAMPLE_XDM_STATE_DATA = new HashMap<String, Object>();
        SAMPLE_XDM_STATE_DATA.put("xdmStateKey", "xdmStateValue");

        when(mockApplication.getSharedPreferences(
                        AssuranceConstants.DataStoreKeys.DATASTORE_NAME, Context.MODE_PRIVATE))
                .thenReturn(mockSharedPreference);
        when(mockSharedPreference.edit()).thenReturn(mockSharedPreferenceEditor);
        assuranceStateManager = new AssuranceStateManager(mockApi, mockApplication);
    }

    @Test
    public void test_ShareAssuranceState() {
        // prepare
        final Event lastSDKEvent = sampleEvent();
        final ArgumentCaptor<Map<String, Object>> sharedStateCaptor =
                ArgumentCaptor.forClass(Map.class);
        final ArgumentCaptor<Event> lastEventCaptor = ArgumentCaptor.forClass(Event.class);
        // simulate last SDK event
        assuranceStateManager.onSDKEvent(lastSDKEvent);

        // test
        assuranceStateManager.shareAssuranceSharedState("SampleSessionId");

        // verify
        verify(mockApi, times(1))
                .createSharedState(sharedStateCaptor.capture(), lastEventCaptor.capture());
        assertEquals(
                "SampleSessionId",
                sharedStateCaptor
                        .getValue()
                        .get(AssuranceTestConstants.SharedStateKeys.ASSURANCE_STATE_SESSION_ID));
        assertNotNull(
                sharedStateCaptor
                        .getValue()
                        .get(AssuranceTestConstants.SharedStateKeys.ASSURANCE_STATE_CLIENT_ID));
        assertNotNull(
                sharedStateCaptor
                        .getValue()
                        .get(
                                AssuranceTestConstants.SharedStateKeys
                                        .ASSURANCE_STATE_INTEGRATION_ID));
        assertEquals(lastSDKEvent, lastEventCaptor.getValue());
    }

    @Test
    public void test_shareAssuranceSharedState_emptySessionID() {
        // prepare
        final Event lastSDKEvent = sampleEvent();
        final ArgumentCaptor<Map<String, Object>> sharedStateCaptor =
                ArgumentCaptor.forClass(Map.class);
        final ArgumentCaptor<Event> lastEventCaptor = ArgumentCaptor.forClass(Event.class);
        // simulate last SDK event
        assuranceStateManager.onSDKEvent(lastSDKEvent);

        // test
        assuranceStateManager.shareAssuranceSharedState("");

        // verify
        verify(mockApi, times(1))
                .createSharedState(sharedStateCaptor.capture(), lastEventCaptor.capture());
        assertNull(
                sharedStateCaptor
                        .getValue()
                        .get(AssuranceTestConstants.SharedStateKeys.ASSURANCE_STATE_SESSION_ID));
        assertNull(
                sharedStateCaptor
                        .getValue()
                        .get(
                                AssuranceTestConstants.SharedStateKeys
                                        .ASSURANCE_STATE_INTEGRATION_ID));
    }

    @Test
    public void test_clearAssuranceState() {
        final ArgumentCaptor<Map<String, Object>> sharedStateCaptor =
                ArgumentCaptor.forClass(Map.class);
        final ArgumentCaptor<Event> lastEventCaptor = ArgumentCaptor.forClass(Event.class);

        // simulate active session id
        assuranceStateManager.shareAssuranceSharedState("SampleSessionId");
        reset(mockApi);

        // test
        assuranceStateManager.clearAssuranceSharedState();

        // verify
        verify(mockApi, times(1))
                .createSharedState(sharedStateCaptor.capture(), lastEventCaptor.capture());
        assertEquals(new HashMap<>(), sharedStateCaptor.getValue());
        assertNull(lastEventCaptor.getValue());
        assertNull(assuranceStateManager.getSessionId());
    }

    @Test
    public void test_geURLEncodedOrgId_validOrgID() {
        // prepare
        setConfigurationSharedStateWithOrgId("B974622245B1A30A490D4D@AdobeOrg");
        // test & verify
        assertEquals("B974622245B1A30A490D4D%40AdobeOrg", assuranceStateManager.getOrgId(true));
    }

    @Test
    public void test_geURLEncodedOrgId_emptyOrgID() {
        // prepare
        setConfigurationSharedStateWithOrgId("");
        // test & verify
        assertEquals("", assuranceStateManager.getOrgId(true));
    }

    @Test
    public void test_geURLEncodedOrgId_nonStringOrgID() {
        // prepare
        setConfigurationSharedStateWithOrgId(423424);

        // test & verify
        assertEquals("", assuranceStateManager.getOrgId(true));
    }

    @Test
    public void test_GetAllExtensionStateData() throws Exception {
        // setup
        String jsonString =
                "{\n"
                        + "  \"extensions\": {\n"
                        + "    \"com.adobe.module.configuration\": {\n"
                        + "      \"version\": \"1.8.0\",\n"
                        + "      \"friendlyName\": \"Configuration\"\n"
                        + "    },\n"
                        + "    \"com.adobe.edge.consent\": {\n"
                        + "      \"version\": \"1.0.0\"\n"
                        + // not providing friendly name for consent
                        "    },\n"
                        + "    \"com.adobe.module.places\": {\n"
                        + "      \"version\": \"1.0.0\",\n"
                        + "      \"friendlyName\": \"com.adobe.module.places\"\n"
                        + "    }\n"
                        + "  },\n"
                        + "  \"version\": \"1.8.0\"\n"
                        + "}";
        final Gson gson = new Gson();
        final Map extensionDetails = gson.fromJson(jsonString, Map.class);

        doAnswer(
                        new Answer<SharedStateResult>() {
                            @Override
                            public SharedStateResult answer(InvocationOnMock invocation)
                                    throws Throwable {
                                String extension = invocation.getArgument(0, String.class);

                                if ("com.adobe.module.places".equals(extension)) {
                                    return null;
                                } else if ("com.adobe.module.configuration".equals(extension)) {
                                    return new SharedStateResult(
                                            SharedStateStatus.SET, SAMPLE_STATE_DATA);
                                } else if ("com.adobe.module.eventhub".equals(extension)) {
                                    return new SharedStateResult(
                                            SharedStateStatus.SET, extensionDetails);
                                } else {
                                    return null;
                                }
                            }
                        })
                .when(mockApi)
                .getSharedState(
                        any(String.class),
                        nullable(Event.class),
                        any(Boolean.class),
                        any(SharedStateResolution.class));

        doAnswer(
                        new Answer<SharedStateResult>() {
                            @Override
                            public SharedStateResult answer(InvocationOnMock invocation)
                                    throws Throwable {
                                String extension = invocation.getArgument(0, String.class);

                                if ("com.adobe.edge.consent".equals(extension)) {
                                    return new SharedStateResult(
                                            SharedStateStatus.SET, SAMPLE_XDM_STATE_DATA);
                                } else {
                                    return null;
                                }
                            }
                        })
                .when(mockApi)
                .getXDMSharedState(
                        any(String.class),
                        nullable(Event.class),
                        any(Boolean.class),
                        any(SharedStateResolution.class));

        // test
        List<AssuranceEvent> stateEvents = assuranceStateManager.getAllExtensionStateData();
        assertEquals(3, stateEvents.size());

        // verify state event 1  : EventHub state
        AssuranceEvent eventHubStateEvent = stateEvents.get(0);
        assertEquals(AssuranceConstants.AssuranceEventType.GENERIC, eventHubStateEvent.type);
        assertEquals(
                "EventHub State",
                eventHubStateEvent.payload.get(
                        AssuranceConstants.GenericEventPayloadKey.ACP_EXTENSION_EVENT_NAME));
        assertEquals(
                EventType.HUB,
                eventHubStateEvent.payload.get(
                        AssuranceConstants.GenericEventPayloadKey.ACP_EXTENSION_EVENT_TYPE));
        assertEquals(
                EventSource.SHARED_STATE,
                eventHubStateEvent.payload.get(
                        AssuranceConstants.GenericEventPayloadKey.ACP_EXTENSION_EVENT_SOURCE));
        assertEquals(
                "com.adobe.module.eventhub",
                ((Map)
                                eventHubStateEvent.payload.get(
                                        AssuranceConstants.GenericEventPayloadKey
                                                .ACP_EXTENSION_EVENT_DATA))
                        .get("stateowner"));
        assertEquals(
                extensionDetails,
                ((Map) eventHubStateEvent.payload.get(AssuranceConstants.PayloadDataKeys.METADATA))
                        .get("state.data"));

        // verify state event 2  : Configuration with regular shared state
        final AssuranceEvent configurationStateEvent = stateEvents.get(1);
        assertEquals(
                "com.adobe.module.configuration",
                ((Map)
                                configurationStateEvent.payload.get(
                                        AssuranceConstants.GenericEventPayloadKey
                                                .ACP_EXTENSION_EVENT_DATA))
                        .get("stateowner"));
        assertEquals(
                SAMPLE_STATE_DATA,
                ((Map)
                                configurationStateEvent.payload.get(
                                        AssuranceConstants.PayloadDataKeys.METADATA))
                        .get("state.data"));

        // verify state event 3  : Consent with XDM shared state
        final AssuranceEvent consentStateEvent = stateEvents.get(2);
        assertEquals(
                "com.adobe.edge.consent",
                ((Map)
                                consentStateEvent.payload.get(
                                        AssuranceConstants.GenericEventPayloadKey
                                                .ACP_EXTENSION_EVENT_DATA))
                        .get("stateowner"));
        assertEquals(
                SAMPLE_XDM_STATE_DATA,
                ((Map) consentStateEvent.payload.get(AssuranceConstants.PayloadDataKeys.METADATA))
                        .get("xdm.state.data"));
    }

    @Test
    public void test_getAllExtensionStateData_noRegisteredExtensions() throws Exception {
        // setup
        Map extensionDetails = Collections.EMPTY_MAP;
        SharedStateResult res = new SharedStateResult(SharedStateStatus.SET, extensionDetails);
        when(mockApi.getSharedState(
                        eq(AssuranceConstants.SDKSharedStateName.EVENTHUB),
                        any(Event.class),
                        any(Boolean.class),
                        any(SharedStateResolution.class)))
                .thenReturn(res);

        // test
        List<AssuranceEvent> stateEvents = assuranceStateManager.getAllExtensionStateData();
        assertEquals(0, stateEvents.size());
    }

    private Event sampleEvent() {
        return new Event.Builder("Mars landing event", EventType.ACQUISITION, EventSource.OS)
                .build();
    }

    private void setConfigurationSharedStateWithOrgId(final Object orgID) {
        final Map<String, Object> configSharedState = new HashMap<>();

        if (orgID != null) {
            configSharedState.put(AssuranceTestConstants.SDKConfigurationKey.ORG_ID, orgID);
        }

        SharedStateResult res = new SharedStateResult(SharedStateStatus.SET, configSharedState);
        doReturn(res)
                .when(mockApi)
                .getSharedState(
                        AssuranceTestConstants.SDKSharedStateName.CONFIGURATION,
                        null,
                        false,
                        SharedStateResolution.ANY);
    }
}

File: code/assurance/src/test/java/com/adobe/marketing/mobile/assurance/AssuranceStateManagerTest.kt
/*
  Copyright 2024 Adobe. All rights reserved.
  This file is licensed to you under the Apache License, Version 2.0 (the "License");
  you may not use this file except in compliance with the License. You may obtain a copy
  of the License at http://www.apache.org/licenses/LICENSE-2.0
  Unless required by applicable law or agreed to in writing, software distributed under
  the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR REPRESENTATIONS
  OF ANY KIND, either express or implied. See the License for the specific language
  governing permissions and limitations under the License.
*/

package com.adobe.marketing.mobile.assurance

import com.adobe.marketing.mobile.EventSource
import com.adobe.marketing.mobile.EventType
import com.adobe.marketing.mobile.ExtensionApi
import com.adobe.marketing.mobile.SharedStateResolution
import com.adobe.marketing.mobile.SharedStateResult
import com.adobe.marketing.mobile.SharedStateStatus
import com.adobe.marketing.mobile.util.JSONUtils
import org.json.JSONArray
import org.json.JSONObject
import org.junit.Before
import org.junit.Test
import org.mockito.Mock
import org.mockito.Mockito.`when`
import org.mockito.MockitoAnnotations
import org.mockito.kotlin.any
import org.mockito.kotlin.anyOrNull
import org.mockito.kotlin.doAnswer
import org.mockito.kotlin.doReturn
import org.mockito.kotlin.eq
import org.mockito.kotlin.verify
import kotlin.test.assertEquals

class AssuranceStateManagerTest {

    companion object {
        const val CONFIGURATION_EXTENSION_STATE_NAME = "com.adobe.module.configuration"
        const val EDGE_EXTENSION_STATE_NAME = "com.adobe.edge"
        const val EDGE_IDENTITY_EXTENSION_STATE_NAME = "com.adobe.edge.identity"
        const val EDGE_CONSENT_EXTENSION_STATE_NAME = "com.adobe.module.consent"
        const val EVENT_HUB_EXTENSION_STATE_NAME = "com.adobe.module.eventhub"

        private val extensionDetails =
            """
            {
              "extensions": {
                "com.adobe.module.configuration": {
                  "version": "3.0.0",
                  "friendlyName": "Configuration"
                },
                "com.adobe.edge": {
                  "version": "3.0.0",
                  "friendlyName": "Edge"
                },
                "com.adobe.edge.identity": {
                  "version": "3.0.0",
                  "friendlyName": "Edge Identity"
                },
                "com.adobe.module.consent": {
                  "version": "3.0.0",
                  "friendlyName": "Edge Consent"
                }
              },
              "wrapper": {
                "type": "N",
                "friendlyName": "None"
              },
              "version": "3.0.0"
            }
            """.trimIndent()

        private val mockSharedStates: Map<String, SharedStateResult> = mutableMapOf(
            CONFIGURATION_EXTENSION_STATE_NAME to SharedStateResult(
                SharedStateStatus.SET,
                mutableMapOf<String, Any?>("orgId" to "orgId")
            ),
            EDGE_EXTENSION_STATE_NAME to SharedStateResult(
                SharedStateStatus.SET,
                mutableMapOf<String, Any?>("edgeKey" to "edgeValue")
            ),
            EDGE_IDENTITY_EXTENSION_STATE_NAME to SharedStateResult(
                SharedStateStatus.SET,
                mutableMapOf<String, Any?>("edgeIdentityKey" to "edgeIdentityValue")
            ),
            EVENT_HUB_EXTENSION_STATE_NAME to SharedStateResult(
                SharedStateStatus.SET,
                JSONUtils.toMap(JSONObject(extensionDetails))
            )
        )

        private val mockXDMSharedStates: Map<String, SharedStateResult> = mutableMapOf(
            EDGE_CONSENT_EXTENSION_STATE_NAME to SharedStateResult(
                SharedStateStatus.SET,
                mutableMapOf<String, Any?>("consentKey" to "consentValue")
            )
        )

        val expectedEventHubStateEvent = mapOf<String, Any?>(
            AssuranceConstants.GenericEventPayloadKey.ACP_EXTENSION_EVENT_NAME to "EventHub State",
            AssuranceConstants.GenericEventPayloadKey.ACP_EXTENSION_EVENT_TYPE to EventType.HUB,
            AssuranceConstants.GenericEventPayloadKey.ACP_EXTENSION_EVENT_SOURCE to EventSource.SHARED_STATE,
            AssuranceConstants.GenericEventPayloadKey.ACP_EXTENSION_EVENT_DATA to mapOf(
                AssuranceConstants.SDKEventDataKey.STATE_OWNER to EVENT_HUB_EXTENSION_STATE_NAME,
            ),
            AssuranceConstants.PayloadDataKeys.METADATA to mapOf<String, Any?>(
                AssuranceConstants.PayloadDataKeys.STATE_DATA to mockSharedStates.getValue(
                    EVENT_HUB_EXTENSION_STATE_NAME
                ).value
            )
        )

        val expectedConfigurationStateEvent = mapOf<String, Any?>(
            AssuranceConstants.GenericEventPayloadKey.ACP_EXTENSION_EVENT_NAME to "Configuration State",
            AssuranceConstants.GenericEventPayloadKey.ACP_EXTENSION_EVENT_TYPE to EventType.HUB,
            AssuranceConstants.GenericEventPayloadKey.ACP_EXTENSION_EVENT_SOURCE to EventSource.SHARED_STATE,
            AssuranceConstants.GenericEventPayloadKey.ACP_EXTENSION_EVENT_DATA to mapOf(
                AssuranceConstants.SDKEventDataKey.STATE_OWNER to CONFIGURATION_EXTENSION_STATE_NAME,
            ),
            AssuranceConstants.PayloadDataKeys.METADATA to mapOf<String, Any?>(
                AssuranceConstants.PayloadDataKeys.STATE_DATA to mockSharedStates.getValue(
                    CONFIGURATION_EXTENSION_STATE_NAME
                ).value
            )
        )

        val expectedEdgeStateEvent = mapOf<String, Any?>(
            AssuranceConstants.GenericEventPayloadKey.ACP_EXTENSION_EVENT_NAME to "Edge State",
            AssuranceConstants.GenericEventPayloadKey.ACP_EXTENSION_EVENT_TYPE to EventType.HUB,
            AssuranceConstants.GenericEventPayloadKey.ACP_EXTENSION_EVENT_SOURCE to EventSource.SHARED_STATE,
            AssuranceConstants.GenericEventPayloadKey.ACP_EXTENSION_EVENT_DATA to mapOf(
                AssuranceConstants.SDKEventDataKey.STATE_OWNER to EDGE_EXTENSION_STATE_NAME,
            ),
            AssuranceConstants.PayloadDataKeys.METADATA to mapOf<String, Any?>(
                AssuranceConstants.PayloadDataKeys.STATE_DATA to mockSharedStates.getValue(
                    EDGE_EXTENSION_STATE_NAME
                ).value
            )
        )

        val expectedEdgeIdentityStateEvent = mapOf<String, Any?>(
            AssuranceConstants.GenericEventPayloadKey.ACP_EXTENSION_EVENT_NAME to "Edge Identity State",
            AssuranceConstants.GenericEventPayloadKey.ACP_EXTENSION_EVENT_TYPE to EventType.HUB,
            AssuranceConstants.GenericEventPayloadKey.ACP_EXTENSION_EVENT_SOURCE to EventSource.SHARED_STATE,
            AssuranceConstants.GenericEventPayloadKey.ACP_EXTENSION_EVENT_DATA to mapOf(
                AssuranceConstants.SDKEventDataKey.STATE_OWNER to EDGE_IDENTITY_EXTENSION_STATE_NAME,
            ),
            AssuranceConstants.PayloadDataKeys.METADATA to mapOf<String, Any?>(
                AssuranceConstants.PayloadDataKeys.STATE_DATA to mockSharedStates.getValue(
                    EDGE_IDENTITY_EXTENSION_STATE_NAME
                ).value
            )
        )

        val expectedConsentStateEvent = mapOf<String, Any?>(
            AssuranceConstants.GenericEventPayloadKey.ACP_EXTENSION_EVENT_NAME to "Edge Consent State",
            AssuranceConstants.GenericEventPayloadKey.ACP_EXTENSION_EVENT_TYPE to EventType.HUB,
            AssuranceConstants.GenericEventPayloadKey.ACP_EXTENSION_EVENT_SOURCE to EventSource.SHARED_STATE,
            AssuranceConstants.GenericEventPayloadKey.ACP_EXTENSION_EVENT_DATA to mapOf(
                AssuranceConstants.SDKEventDataKey.STATE_OWNER to EDGE_CONSENT_EXTENSION_STATE_NAME,
            ),
            AssuranceConstants.PayloadDataKeys.METADATA to mapOf<String, Any?>(
                AssuranceConstants.PayloadDataKeys.XDM_STATE_DATA to mockXDMSharedStates.getValue(
                    EDGE_CONSENT_EXTENSION_STATE_NAME
                ).value
            )
        )
    }

    @Mock
    private lateinit var mockAssuranceSharedStateManager: AssuranceSharedStateManager

    @Mock
    private lateinit var mockExtensionApi: ExtensionApi

    private lateinit var assuranceStateManager: AssuranceStateManager

    @Before
    fun setup() {
        MockitoAnnotations.openMocks(this)

        assuranceStateManager =
            AssuranceStateManager(mockExtensionApi, mockAssuranceSharedStateManager)
    }

    @Test
    fun `Test #getAllExtensionStateData() with Standard and XDM states`() {
        `when`(
            mockExtensionApi.getSharedState(
                any(),
                eq(null),
                eq(false),
                eq(SharedStateResolution.ANY)
            )
        ).doAnswer {
            val extensionName = it.arguments[0] as String
            mockSharedStates[extensionName]
        }

        `when`(
            mockExtensionApi.getXDMSharedState(
                any(),
                eq(null),
                eq(false),
                eq(SharedStateResolution.ANY)
            )
        ).doAnswer {
            val extensionName = it.arguments[0] as String
            mockXDMSharedStates[extensionName]
        }

        // Test
        val stateEvents = assuranceStateManager.getAllExtensionStateData()

        // Verify the number of state events are as expected
        assertEquals(5, stateEvents.size)

        // Prepare an expected list of state events to compare with the actual state events
        val expectedStateMaps = mutableListOf(
            expectedEventHubStateEvent.toMutableMap(),
            expectedConfigurationStateEvent.toMutableMap(),
            expectedEdgeStateEvent.toMutableMap(),
            expectedEdgeIdentityStateEvent.toMutableMap(),
            expectedConsentStateEvent.toMutableMap()
        )

        val actualStateMaps = stateEvents.map { it.payload }
        assertEquals(JSONArray(expectedStateMaps).toString(), JSONArray(actualStateMaps).toString())
    }

    @Test
    fun `Test #getAllExtensionStateData() with no XDM shared states`() {
        // Setup
        `when`(
            mockExtensionApi.getSharedState(
                any(),
                eq(null),
                eq(false),
                eq(SharedStateResolution.ANY)
            )
        ).doAnswer {
            val extensionName = it.arguments[0] as String
            mockSharedStates[extensionName]
        }

        // Do not return any XDM shared state
        `when`(
            mockExtensionApi.getXDMSharedState(
                any(),
                eq(null),
                eq(false),
                eq(SharedStateResolution.ANY)
            )
        ).doAnswer {
            SharedStateResult(SharedStateStatus.NONE, null)
        }

        // Test
        val stateEvents = assuranceStateManager.getAllExtensionStateData()

        // Verify the number of state events are as expected
        assertEquals(4, stateEvents.size)

        // Prepare an expected list of state events to compare with the actual state events
        val expectedStateMaps = mutableListOf(
            expectedEventHubStateEvent.toMutableMap(),
            expectedConfigurationStateEvent.toMutableMap(),
            expectedEdgeStateEvent.toMutableMap(),
            expectedEdgeIdentityStateEvent.toMutableMap()
            // No Edge Consent state event as there is no XDM shared state
        )

        val actualStateMaps = stateEvents.map { it.payload }
        assertEquals(JSONArray(expectedStateMaps).toString(), JSONArray(actualStateMaps).toString())
    }

    @Test
    fun `Test #getOrgId with url encoding`() {
        val orgId = "B974622245B1A30A490D4D@AdobeOrg"
        setConfigurationSharedStateWithOrgId(orgId)

        val result = assuranceStateManager.getOrgId(true)

        assertEquals("B974622245B1A30A490D4D%40AdobeOrg", result)
    }

    @Test
    fun `Test #getOrgId without url encoding`() {
        val orgId = "B974622245B1A30A490D4D@AdobeOrg"
        setConfigurationSharedStateWithOrgId(orgId)

        val result = assuranceStateManager.getOrgId(false)

        assertEquals(orgId, result)
    }

    @Test
    fun `Test #getOrgId with empty orgId`() {
        val orgId = ""
        setConfigurationSharedStateWithOrgId(orgId)

        val result = assuranceStateManager.getOrgId(false)

        assertEquals(orgId, result)
    }

    @Test
    fun `Test #getOrgId with no configuration state`() {
        `when`(mockExtensionApi.getSharedState(AssuranceConstants.SDKSharedStateName.CONFIGURATION, null, false, SharedStateResolution.ANY))
            .thenReturn(SharedStateResult(SharedStateStatus.NONE, null))

        val result = assuranceStateManager.getOrgId(false)

        assertEquals("", result)
    }

    @Test
    fun `Test #shareAssuranceSharedState() with null SessionId`() {
        `when`(mockAssuranceSharedStateManager.assuranceSharedState).thenReturn(AssuranceSharedState("clientId", "sessionId"))

        assuranceStateManager.shareAssuranceSharedState(null)

        verify(mockAssuranceSharedStateManager).setSessionId(null)
        verify(mockExtensionApi).createSharedState(anyOrNull(), anyOrNull())
    }

    @Test
    fun `Test #shareAssuranceSharedState() with empty SessionId`() {
        `when`(mockAssuranceSharedStateManager.assuranceSharedState).thenReturn(AssuranceSharedState("clientId", "sessionId"))

        assuranceStateManager.shareAssuranceSharedState("")

        verify(mockAssuranceSharedStateManager).setSessionId("")
        verify(mockExtensionApi).createSharedState(anyOrNull(), anyOrNull())
    }

    @Test
    fun `Test #clearAssuranceSharedState() with non-empty SessionId`() {
        assuranceStateManager.clearAssuranceSharedState()

        verify(mockAssuranceSharedStateManager).setSessionId(null)
        verify(mockExtensionApi).createSharedState(mapOf(), null)
    }

    private fun setConfigurationSharedStateWithOrgId(orgId: String) {
        val configurationSharedState = mutableMapOf<String, Any?>()

        orgId.let {
            configurationSharedState[AssuranceConstants.SDKConfigurationKey.ORG_ID] = it
        }

        val sharedStateResult = SharedStateResult(SharedStateStatus.SET, configurationSharedState)
        doReturn(sharedStateResult).`when`(mockExtensionApi).getSharedState(
            AssuranceConstants.SDKSharedStateName.CONFIGURATION,
            null,
            false,
            SharedStateResolution.ANY
        )
    }
}
